enclave/sgx/verifier.c

```diff
@@ -0,0 +1,308 @@
+    OE_UNUSED(p_qe_identity);
+    OE_UNUSED(qe_identity_size);
+    OE_UNUSED(p_qe_identity_issuer_chain);
+    OE_UNUSED(qe_identity_issuer_chain_size);
+
+
+    if (_retval)
+        *_retval = OE_UNSUPPORTED;
+
+    return OE_UNSUPPORTED;
+}
+OE_WEAK_ALIAS(
+    _oe_verify_quote_ocall, oe_verify_quote_ocall);
+#endif
+
+
+oe_result_t oe_verify_qve_report_and_identity(
+        const uint8_t* p_quote,
+        uint32_t quote_size,
+        const sgx_ql_qe_report_info_t* p_qve_report_info,
+        time_t expiration_check_date,
+        uint32_t collateral_expiration_status,
+        uint32_t quote_verification_result,
+        const uint8_t *p_supplemental_data,
+        uint32_t supplemental_data_size,
+        uint16_t qve_isvsvn_threshold)
+{
+    oe_result_t result = OE_UNEXPECTED;
+
+    oe_sha256_context_t sha_handle = { 0 };
+    OE_SHA256 report_data_hash = { 0 };
+    sgx_report_data_t report_data = { 0 };
+
+    if (p_quote == NULL ||
+            p_qve_report_info == NULL ||
+            sizeof(*p_qve_report_info) != sizeof(sgx_ql_qe_report_info_t) ||
+            !oe_is_within_enclave(p_quote, quote_size) ||
+            !oe_is_within_enclave(p_qve_report_info, sizeof(sgx_ql_qe_report_info_t)) ||
+            (p_supplemental_data == NULL && supplemental_data_size != 0) ||
+            (p_supplemental_data != NULL && supplemental_data_size == 0))
+        return OE_INVALID_PARAMETER;
+
+    if (p_supplemental_data && supplemental_data_size > 0) {
+        if (!oe_is_within_enclave(p_supplemental_data, supplemental_data_size)) {
+            return OE_INVALID_PARAMETER;
+        }
+    }
+
+    //Defense in depth, threshold must be greater or equal to 3
+    //
+    if (qve_isvsvn_threshold < LEAST_QVE_ISVSVN)
+        return OE_QUOTE_ENCLAVE_IDENTITY_VERIFICATION_FAILED;
+
+    // Add oe report header info for QvE report
+    oe_report_header_t* p_oe_qve_report = (oe_report_header_t*) malloc(
+                                            sizeof(oe_report_header_t) +
+                                            sizeof(sgx_report_t));
+    memset(p_oe_qve_report, 0, sizeof(oe_report_header_t) + sizeof(sgx_report_t));
+
+    uint8_t* p_offset = (uint8_t*)p_oe_qve_report + sizeof(oe_report_header_t);
+
+    if (p_oe_qve_report == NULL) {
+        OE_RAISE_MSG(
+            OE_OUT_OF_MEMORY,
+            "Out of memory when creating OE reort header buffer during QvE ID verification",
+            NULL);
+    }
+
+    p_oe_qve_report->version = OE_REPORT_HEADER_VERSION;
+    p_oe_qve_report->report_type = OE_REPORT_TYPE_SGX_LOCAL;
+    p_oe_qve_report->report_size = sizeof(sgx_report_t);
+    
+    OE_CHECK(oe_memcpy_s(
+        p_offset,
+        sizeof(sgx_report_t),
+        &(p_qve_report_info->qe_report),
+        sizeof(sgx_report_t)
+    ));
+    
+    const sgx_report_t *p_qve_report = &(p_qve_report_info->qe_report);
+
+    //verify QvE report
+    OE_CHECK_MSG(oe_verify_report((uint8_t*)p_oe_qve_report, 
```

Comment:
Can we call function to verify `p_qve_report_info->qe_report` without constructing an `oe_report_t`? I think `oe_verify_raw_sgx_report()` is the internal function for that purpose.

Reply:
Yes, `oe_verify_raw_sgx_report()` should work, will update.